api/app.py

```diff
@@ -41,39 +51,130 @@
 
 @app.route('/add', methods=['POST'])
 def add():
+    # extract arguments from post request
     try:
         text = request.form['text']
         category = request.form['category']
+        id_token = request.form['id_token']
     except KeyError:
-        return "Error: must specify the 'text' and 'category' keys when making a POST to /add"
+        return "Error: must specify the 'text', 'category', and 'id_token' keys when making a POST to /add"
+
+    try:
+        # Verify the ID token while checking if the token is revoked by
+        # passing check_revoked=True
+        decoded_token = auth.verify_id_token(id_token, check_revoked=True)
+        # Token is valid and not revoked
+        uid = decoded_token['uid']
+    except ValueError:
+        return "Error: token malformed"
+    except auth.AuthError as exc:
+        if exc.code == 'ID_TOKEN_REVOKED':
+            return "Error: token revoked; user must reauthenticate or sign out"
+        else:
+            return "Error: token invalid"
+
+    # get the actual user from the uid
+    user = auth.get_user(uid)
+
+    # control access using custom claim
+    print(user.custom_claims)
+    if user.custom_claims is None or user.custom_claims.get('admin') is not True:
+        # Prevent access to admin resource.
+        return "Error: user is not allowed to access this resource"
+
+    # verify prediction pre-requisites
     if text in data:
         return "Error: data store already contains text"
+
+    # add given text to in-memory data store
     data.append(text)
     if category not in cats:
         cats.append(category)
     target.append(cats.index(category))
+
+    # build classifier if pre-requisites satisfied
     # if len(cats) >= 2:
         # requires enough categories to make a meaningful prediction
         # text_clf.fit(data, target)
     if len(cats) >= 2 and all(i >= 3 for i in Counter(target).values()):
         # requires enough data to split into training, test, and validation sets
         gs_clf.fit(data, target)
+
+    # return value is arbitrary
     return "added " + text + " to category " + category + " which is index " + str(cats.index(category))
 
 @app.route('/predict', methods=['GET'])
 def predict():
+    # extract arguments from get request
     try:
         text = request.args['text']
+        id_token = request.args['id_token']
     except KeyError:
-        return "Error: must specify the 'text' arg when making a GET to /predict"
+        return "Error: must specify the 'text' and 'id_token' args when making a GET to /predict"
+
+    try:
+        # Verify the ID token while checking if the token is revoked by
+        # passing check_revoked=True
+        decoded_token = auth.verify_id_token(id_token, check_revoked=True)
```

Comment:
When we move to having the client establish a session with the server, we should be able to verify the token at the beginning of the session and not in each request, correct?

Reply:
Yeah, I think that's right for socket connections; just once during the socket negotiation.

If we keep using HTTP requests however, we need tokens every time because HTTP requests are stateless. We could try to get around this with session cookies or something, but iirc there are security implications and that verifying with every request is easier to keep secure.